Source/Dafny/Compilers/Compiler-Csharp.cs

```diff
@@ -3207,5 +3291,259 @@ private class CSharpCompilationResult {
+    ///         | SKOLEMS, SKOLEMS
+    /// 
+    /// </summary>
+    private class MockWriter {
+
+      private readonly CsharpCompiler compiler;
+      // maps identifier names to the names of corresponding mocks:
+      private Dictionary<string, string> mockNames = new();
+      // associates a skolem variable with the lambda passed to argument matcher
+      private List<Tuple<IVariable, string>> bounds = new();
+
+      public MockWriter(CsharpCompiler compiler) {
+        this.compiler = compiler;
+      }
+
+      /// <summary>
+      /// Create a body for a method returning a fresh instance of an object 
+      /// </summary>
+      public ConcreteSyntaxTree CreateFreshMethod(Method m,
+        ConcreteSyntaxTree wr) {
+        var keywords = Keywords(true, true);
+        var returnType = compiler.GetTargetReturnTypeReplacement(m, wr);
+        wr.FormatLine($"{keywords}{returnType} {compiler.IdName(m)}() {{");
+        // Exploit the fact that compiler creates zero-arguments constructors:
+        wr.FormatLine($"return new {returnType}();");
+        wr.WriteLine("}");
+        return wr;
+      }
+
+      /// <summary>
+      /// Create a body of a method that mocks one or more objects
+      /// </summary>
+      public ConcreteSyntaxTree CreateMockMethod(Method m,
+        List<TypeArgumentInstantiation> typeArgs, bool createBody,
+        ConcreteSyntaxTree wr, bool forBodyInheritance, bool lookasideBody) {
+
+        // The following few lines are identical to those in CreateMethod above:
+        var customReceiver = createBody &&
+                             !forBodyInheritance &&
+                             compiler.NeedsCustomReceiver(m);
+        var keywords = Keywords(true, true);
+        var returnType = compiler.GetTargetReturnTypeReplacement(m, wr);
+        var typeParameters = compiler.TypeParameters(TypeArgumentInstantiation.
+          ToFormals(compiler.ForTypeParameters(typeArgs, m, lookasideBody)));
+        var parameters = compiler
+          .GetMethodParameters(m, typeArgs, lookasideBody, customReceiver, returnType);
+
+        // Out parameters cannot be used inside lambda expressions in Csharp
+        // but the mocked objects may appear in lambda expressions during
+        // mocking (e.g. two objects may cross-reference each other).
+        // The solution is to rename the out parameters.
+        var parameterString = parameters.ToString();
+        var retNames = new Dictionary<string, string>();
+        foreach (var o in m.Outs) {
+          retNames[o.CompileName] = FreshId(o.CompileName + "Return");
+          parameterString = Regex.Replace(parameterString,
+            $"(^|[^a-zA-Z0-9_]){o.CompileName}([^a-zA-Z0-9_]|$)",
+            "$1" + retNames[o.CompileName] + "$2");
+        }
+        wr.FormatLine($"{keywords}{returnType} {compiler.IdName(m)}{typeParameters}({parameterString}) {{");
+
+        // Initialize the mocks:
+        mockNames = new Dictionary<string, string>();
```

Comment:
Consider writing 

```csharp
mockNames = m.Outs.ToDictionary(o => o.CompileName, o => compiler.idGenerator.FreshId(o.CompileName + "Mock"));
foreach(var (methodName, mockName) in mockNames) {
  wr.FormatLine($"var {mockName} = new Mock<{compiler.TypeName(o.Type, wr, o.tok)}>();");
  wr.FormatLine($"var {methodName} = {mockName}.Object;");
}
```

to separate the two concerns.

Consider doing something similar for the initialization of `retNames`.

Reply:
Done!